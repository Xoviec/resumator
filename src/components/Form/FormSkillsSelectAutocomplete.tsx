import React, { FunctionComponent } from "react";
import { Autocomplete, AutocompleteChangeReason } from "@material-ui/lab";
import { makeStyles, TextField } from "@material-ui/core";
import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";
import FormSkillsSelectChip from "./FormSkillsSelectChip";

interface Skill {
  name: string;
}

interface FormSkillsSelectPropsAutocomplete {
  label?: string;
  value: Skill[];
  onChange: (skills: Skill[]) => void;
  options: string[];
}

const useStyles = makeStyles({
  autocomplete: {
    // Make sure the input is below the chips.
    "& .MuiAutocomplete-input": {
      width: "inherit",
    },
  },
  textField: {
    "& input": {
      marginTop: "4px",
    },
  },
});

const FormSkillsSelectAutocomplete: FunctionComponent<FormSkillsSelectPropsAutocomplete> =
  ({ label, value, onChange, options }) => {
    const classes = useStyles();
    /**
     * Check if the provided option is currently included in the skills.
     */
    const getOptionSelected = (option: string, skill: string) => option === skill;

    /**
     * Handle adding or deleting a skill through the autocomplete input.
     */
    const handleSkillChange = (
      event: React.ChangeEvent<Record<string, unknown>>,
      inputValue: string[],
      reason: AutocompleteChangeReason
    ) => {
      const skills = inputValue.map((name) => ({ name }));
      onChange(skills);
    };

    /**
     * Handle deleting a skill by clicking the x on the chip.
     */
    const handleSkillDelete = (index: number) =>
      onChange(value.filter((skill, i) => index !== i));

    /**
     * Handle when a skill is being dropped in a new position.
     */
    const handleDrag = (sourceIndex: number, destinationIndex: number) => {
      // Not a full copy, but as we don't edit skills that should be okay.
      const skills = [...value];
      skills.splice(destinationIndex, 0, skills.splice(sourceIndex, 1)[0]);

      onChange(skills);
    };

    return (
      <DndProvider backend={HTML5Backend}>
        <Autocomplete
          fullWidth
          multiple
          disableClearable
          disableCloseOnSelect
          id="skill-list-autocomplete"
          size="small"
          className={classes.autocomplete}
          value={value.map((skill) => skill.name)}
          options={options}
          onChange={handleSkillChange}
          getOptionSelected={getOptionSelected}
          renderInput={(params) => (
            <TextField
              variant="outlined"
              placeholder="Add a library, framework, skill..."
              className={value.length ? classes.textField : undefined}
              label={label}
              {...params}
            />
          )}
          renderTags={(value: string[]) =>
            value.map((skill, index) => (
              // Add a chip for each skill.
              <FormSkillsSelectChip
                key={skill}
                label={skill}
                index={index}
                onDrag={handleDrag}
                onDelete={handleSkillDelete}
              />
            ))
          }
        />
      </DndProvider>
    );
  };

export default FormSkillsSelectAutocomplete;
