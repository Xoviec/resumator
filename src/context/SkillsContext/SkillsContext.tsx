import {
  createContext,
  FunctionComponent,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useCollection } from "react-firebase-hooks/firestore";
import { useFirebaseApp } from "../FirebaseContext";

type SkillsContextType = {
  skillList: string[];
  updateSkillList: (arg: string[]) => Promise<void>;
};

const SkillsContext = createContext<SkillsContextType | undefined>(undefined);

export const useSkillsContext = (): SkillsContextType => {
  const context = useContext(SkillsContext);
  if (context === undefined) {
    throw new Error("Firebase context is used before initialization");
  }

  return context;
};

export const SkillsContextProvider: FunctionComponent = ({ children }) => {
  const { firebase } = useFirebaseApp();

  const [skillList, setSkillList] = useState<string[]>([]);
  const [docId, setDocId] = useState<string>("");
  const [val] = useCollection(firebase.firestore().collection("allSkills"));

  useEffect(() => {
    if (val) {
      setDocId(val.docs[0].id);

      const uniqueSkillList = [
        ...Array.from(new Set(val.docs[0].data().skills)),
      ] as string[];

      setSkillList(uniqueSkillList);
    }
  }, [val]);

  const updateSkillList: SkillsContextType["updateSkillList"] = useMemo(() => {
    return async (newSkillList) => {
      const uniqueNewSkillList = [...Array.from(new Set(newSkillList))];

      // it is very likely for the firebase update to be successful, so it covers most of the cases
      // so we ensure faster user experience for majority of users
      // in the unlikely scenario, it will be updated from firebase anyway
      setSkillList(uniqueNewSkillList);

      await firebase
        .firestore()
        .collection("allSkills")
        .doc(docId)
        .update({ skills: uniqueNewSkillList });
    };
  }, [docId, firebase]);

  const skillsContext: SkillsContextType = useMemo(
    () => ({
      skillList,
      updateSkillList,
    }),
    [skillList, updateSkillList]
  );

  return (
    <SkillsContext.Provider value={skillsContext}>{children}</SkillsContext.Provider>
  );
};
