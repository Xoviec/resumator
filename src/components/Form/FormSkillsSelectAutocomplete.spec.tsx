import { fireEvent, render } from "@testing-library/react";
import { mocked } from "jest-mock";
import { SortableContainer } from "react-sortable-hoc";
import { useSkillsContext } from "../../context/SkillsContext/SkillsContext";
import { FormSkillsSelectAutocomplete } from "./FormSkillsSelectAutocomplete";

jest.mock("../../context/SkillsContext/SkillsContext");
jest.mock("react-sortable-hoc");

beforeEach(jest.clearAllMocks);

describe("FormSkillsSelectAutocomplete", () => {
  it("should render SortableSelect with correct props", async () => {
    const SortableSelect = jest.fn(() => <div>sortable-container</div>);

    mocked(SortableContainer).mockImplementation(() => SortableSelect as any);

    mocked(useSkillsContext).mockImplementation(
      () =>
        ({
          skillList: [],
          updateSkillList: () => {},
        } as any)
    );

    render(
      <FormSkillsSelectAutocomplete
        label="test-label"
        value={[{ name: "test-skill1" }]}
        onChange={() => {}}
      />
    );

    expect(SortableSelect).toMatchSnapshot();
  });

  it("should call onChange when sorting ends", async () => {
    const SortableSelect = ({ onSortEnd }: any) => {
      return (
        <button
          onClick={() => {
            onSortEnd({ oldIndex: 0, newIndex: 1 });
          }}
          data-testid="sortable-select"
        >
          sortable-select
        </button>
      );
    };

    mocked(SortableContainer).mockImplementation(() => SortableSelect as any);

    mocked(useSkillsContext).mockImplementation(
      () =>
        ({
          skillList: [],
          updateSkillList: () => {},
        } as any)
    );

    const onChange = jest.fn();

    const { getByTestId } = render(
      <FormSkillsSelectAutocomplete
        label="test-label"
        value={[{ name: "test-skill1" }, { name: "test-skill2" }]}
        onChange={onChange}
      />
    );

    fireEvent.click(getByTestId("sortable-select"));
    expect(onChange).toHaveBeenCalledWith([
      { name: "test-skill2", value: "test-skill2", label: "test-skill2" },
      { name: "test-skill1", value: "test-skill1", label: "test-skill1" },
    ]);
  });

  it("should call onChange when data changes", async () => {
    const SortableSelect = ({ onChange }: any) => {
      return (
        <button
          onClick={() => {
            onChange([
              { name: "test-skill2", value: "test-skill2", label: "test-skill2" },
              { name: "test-skill1", value: "test-skill1", label: "test-skill1" },
            ]);
          }}
          data-testid="sortable-select"
        >
          sortable-select
        </button>
      );
    };

    mocked(SortableContainer).mockImplementation(() => SortableSelect as any);

    mocked(useSkillsContext).mockImplementation(
      () =>
        ({
          skillList: [],
          updateSkillList: () => {},
        } as any)
    );

    const onChange = jest.fn();

    const { getByTestId } = render(
      <FormSkillsSelectAutocomplete
        label="test-label"
        value={[{ name: "test-skill1" }, { name: "test-skill2" }]}
        onChange={onChange}
      />
    );

    fireEvent.click(getByTestId("sortable-select"));
    expect(onChange).toHaveBeenCalledWith([
      { name: "test-skill2", value: "test-skill2", label: "test-skill2" },
      { name: "test-skill1", value: "test-skill1", label: "test-skill1" },
    ]);
  });

  it("should provide options from skillList", async () => {
    const SortableSelect = jest.fn(() => <div>sortable-container</div>);

    mocked(SortableContainer).mockImplementation(() => SortableSelect as any);

    mocked(useSkillsContext).mockImplementation(
      () =>
        ({
          skillList: ["skill1", "skill2"],
          updateSkillList: () => {},
        } as any)
    );

    const onChange = jest.fn();

    render(
      <FormSkillsSelectAutocomplete
        label="test-label"
        value={[{ name: "test-skill1" }, { name: "test-skill2" }]}
        onChange={onChange}
      />
    );

    const sortableSelect: [][] = SortableSelect.mock.calls[0];

    expect(sortableSelect[0]).toMatchObject({
      options: [
        { value: "skill1", label: "skill1", isActive: true },
        { value: "skill2", label: "skill2", isActive: true },
      ],
    });
  });
});
