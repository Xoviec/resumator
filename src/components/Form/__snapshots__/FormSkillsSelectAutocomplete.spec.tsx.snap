// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSkillsSelectAutocomplete should render SortableSelect with correct props 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "axis": "xy",
        "closeMenuOnSelect": false,
        "components": Object {
          "MultiValue": undefined,
          "MultiValueLabel": undefined,
        },
        "distance": 4,
        "getHelperDimensions": [Function],
        "isMulti": true,
        "menuPosition": "fixed",
        "onChange": [Function],
        "onSortEnd": [Function],
        "options": Array [],
        "placeholder": "Add a library, framework, skill...",
        "useDragHandle": true,
        "value": Array [],
      },
      Object {},
    ],
    Array [
      Object {
        "axis": "xy",
        "closeMenuOnSelect": false,
        "components": Object {
          "MultiValue": undefined,
          "MultiValueLabel": undefined,
        },
        "distance": 4,
        "getHelperDimensions": [Function],
        "isMulti": true,
        "menuPosition": "fixed",
        "onChange": [Function],
        "onSortEnd": [Function],
        "options": Array [],
        "placeholder": "Add a library, framework, skill...",
        "useDragHandle": true,
        "value": Array [
          Object {
            "label": "test-skill1",
            "name": "test-skill1",
            "value": "test-skill1",
          },
        ],
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": <div>
        sortable-container
      </div>,
    },
    Object {
      "type": "return",
      "value": <div>
        sortable-container
      </div>,
    },
  ],
}
`;
